roles/lib_openshift/src/class/oadm_router.py

```diff
@@ -0,0 +1,437 @@
+                rfd.write(open(self.config.config_options['cacert_file']['value']).read())
+
+        atexit.register(Utils.cleanup, [router_pem])
+        self.config.config_options['default_cert']['value'] = router_pem
+
+        options = self.config.to_option_list()
+
+        cmd = ['router', self.config.name, '-n', self.config.namespace]
+        cmd.extend(options)
+        cmd.extend(['--dry-run=True', '-o', 'json'])
+
+        results = self.openshift_cmd(cmd, oadm=True, output=True, output_type='json')
+
+        # pylint: disable=no-member
+        if results['returncode'] != 0 and results['results'].has_key('items'):
+            return results
+
+        oc_objects = {'DeploymentConfig': {'obj': None, 'path': None},
+                      'Secret': {'obj': None, 'path': None},
+                      'ServiceAccount': {'obj': None, 'path': None},
+                      'ClusterRoleBinding': {'obj': None, 'path': None},
+                      'Service': {'obj': None, 'path': None},
+                     }
+        # pylint: disable=invalid-sequence-index
+        for res in results['results']['items']:
+            if res['kind'] == 'DeploymentConfig':
+                oc_objects['DeploymentConfig']['obj'] = DeploymentConfig(res)
+            elif res['kind'] == 'Service':
+                oc_objects['Service']['obj'] = Service(res)
+            elif res['kind'] == 'ServiceAccount':
+                oc_objects['ServiceAccount']['obj'] = ServiceAccount(res)
+            elif res['kind'] == 'Secret':
+                oc_objects['Secret']['obj'] = Secret(res)
+            elif res['kind'] == 'ClusterRoleBinding':
+                oc_objects['ClusterRoleBinding']['obj'] = RoleBinding(res)
+
+        # Currently only deploymentconfig needs updating
+        # Verify we got a deploymentconfig
+        if not oc_objects['DeploymentConfig']['obj']:
+            return results
+
+        # results will need to get parsed here and modifications added
+        oc_objects['DeploymentConfig']['obj'] = self.add_modifications(oc_objects['DeploymentConfig']['obj'])
+
+        for oc_type in oc_objects.keys():
+            oc_objects[oc_type]['path'] = Utils.create_tmp_file_from_contents(oc_type, oc_objects[oc_type]['obj'].yaml_dict)
+
+        return oc_objects
+
+    def create(self):
+        '''Create a deploymentconfig '''
+        # generate the objects and prepare for instantiation
+        self.prepare_router()
+
+        results = []
+        for _, oc_data in self.router_prep.items():
+            results.append(self._create(oc_data['path']))
+
+        rval = 0
+        for result in results:
+            if result['returncode'] != 0 and not 'already exist' in result['stderr']:
+                rval = result['returncode']
+
+        return {'returncode': rval, 'results': results}
+
+    def update(self):
+        '''run update for the router.  This performs a delete and then create '''
```

Comment:
> I disagree with the regression mentioned. The current behavior is to do nothing if there are changes. The modules I'm adding here will at least make the comparisons and attempt to reconcile if there are changes.

So, I don't think this is entirely accurate.

That is the case when re-running openshift_hosted. It does seem that I was mistaken about the number of replicas, for some reason I thought we were doing a scale operation to adjust the replica count, but I can see that we actually are not.

For upgrade, we are currently patching the dc of the router/registry to update the required items, so if we handled upgrades using these router/registry modules, it would indeed be a regression from the current behavior. 

> I like the update or patch idea. The difficulty comes in on how do you determine which sections to update. We can discuss this further as I support the idea of updating instead of delete/recreate.

There are a few different ways to tackle it, but if starting with the current implementation then the most expedient approach would depend on whether patch or update is going to be used.

For patch, a method similar to check_def_equal that returns back a minimal datastructure consisting of the `user_def` fields that differ from `result_def` instead of returing True/False. check_def_equal could most likely be updated to wrap this new method, returning False if the patch object is not `{}` or `None`. Since patch would only be called if there is a diff being applied, then the return code of patch could be used to determine success/failure of the patch.

For update, instead of creating a minimal patch you could create a copy of `result_def`. You would also need to strip out read-only fields from metadata (https://kubernetes.io/docs/api-reference/v1/definitions/#_v1_objectmeta) and the object status before applying the update.

> As a side note, the delete/recreate is very fast. It does cause a small window of outage but I think it was probably the quickest and easiest path. We might have also run into some issues with openshift querying or modifying the objects while we were inspecting them. What generally would happen is that the object ids would get incremented and by the time you would attempt a replace it would not accept the version you were adding. Thus, delete / recreate would always work here.

I think you mean resource versions instead of object ids. With the delete/recreate the object uids would change, but update/patch operations would not affect the uids.

Even if the delete/recreate is fast and the outage window is small, I personally would consider this a blocker because the outages will affect the following classes of users:
- Users that are running openshift-ansible periodically for ongoing CM. There was at least one user that was doing this in the past (I'm not sure if they still are), and I know I've helped consultants that have worked on managing their clusters using Ansible Tower and openshift-ansible.
- Users that are expecting blue/green upgrades to incur zero downtime. Since we just delivered this for 1.4/3.4, shipping a regression to this behavior (however small) seems like a step backwards.
- Users expecting certificate-redeploys to incur zero downtime.